src/loader/mod.rs

```diff
@@ -211,20 +215,24 @@ pub fn load_cmdline<M: GuestMemory>(
     guest_addr: GuestAddress,
     cmdline: &Cmdline,
 ) -> Result<()> {
-    let len = cmdline.as_str().len();
-    if len == 0 {
-        return Ok(());
-    }
+    // We need a null terminated string because that's what the Linux
+    // kernel expects when parsing the command line:
+    // https://elixir.bootlin.com/linux/v5.10.139/source/kernel/params.c#L179
+    let cmdline_string = cmdline
+        .as_cstring()
+        .map_err(|_| Error::InvalidCommandLine)?;
+
+    let cmdline_bytes = cmdline_string.as_bytes_with_nul();
 
     let end = guest_addr
-        .checked_add(len as u64 + 1)
-        .ok_or(Error::CommandLineOverflow)?; // Extra for null termination.
+        .checked_add(cmdline_bytes.len() as u64)
+        .ok_or(Error::CommandLineOverflow)?;
```

Comment:
Hmm I just noticed.. so we are returning an error if the first address after the command line area is outside the u64 range. Shouldn't we actually do that for the last address in the command line region?

Reply:
Hmm, looks like we also don't have a test for this corner case.

Reply:
Regarding the tests we have, there is a test that checks the overflow of the memory. Regarding the overflow of that addition ... yes, we don't have one; I'll add this usecase in the existing overflow tes

Reply:
To be honest, the way we compute the value of the end variable seems to be not accurate; with the current implementation, theoretically, we throw an error even if the the cmdline (including null terminator) is within guest memory's range, but the null terminator is in the last byte of the guest (because the next address will be bigger then what guest_mem.last_addr() returns which is the last byte of the guest's memory) - this is a very specific and improbable usecase... I don't even know if it's technically possible to store the cmdline at the end of the guest memory but in the context of this comment, it is worth mentioning this aspect. Am I right? Did I miss something? Should I change`let end = guest_addr
        .checked_add(cmdline_bytes.len() as u64)
        .ok_or(Error::CommandLineOverflow)?;` into `let end = guest_addr
        .checked_add((cmdline_bytes.len() - 1) as u64)
        .ok_or(Error::CommandLineOverflow)?;` ?

Reply:
Leaving this comment unaddressed for further discussions. I'll push the changes for the other comments.

Reply:
> To be honest, the way we compute the value of the end variable seems to be not accurate; with the current implementation, theoretically, we throw an error even if the the cmdline (including null terminator) is within guest memory's range, but the null terminator is in the last byte of the guest (because the next address will be bigger then what guest_mem.last_addr() returns which is the last byte of the guest's memory) - this is a very specific and improbable usecase... I don't even know if it's technically possible to store the cmdline at the end of the guest memory but in the context of this comment, it is worth mentioning this aspect. Am I right? Did I miss something? Should I change`let end = guest_addr.checked_add(cmdline_bytes.len() as u64).ok_or(Error::CommandLineOverflow)?;` into `let end = guest_addr.checked_add((cmdline_bytes.len() - 1) as u64).ok_or(Error::CommandLineOverflow)?;` ?

Yes, this is exactly what I was suggesting.

Reply:
Ah... reading your comment I understood that we should somehow check if the load_cmdline stores the cmdline bytes into a range of addresses that the linux kernel expects to find the cmdline; sorry for my misunderstanding; fixing that right now